### ISSUE TITLE
2.5.3 Label in Name, do we need an explanation of a "visible" label?

### ISSUE THREAD

**Issue by anevins12** [w3c/wcag#3229]
Do we need an explanation of what it means for a label to be "visible" in 2.5.3 Label in Name? It is unclear whether a partially visible label is applicable, such as one that appears only after interaction like a tooltip.

As one of the intended user groups for 2.5.3 is speech users, some auditors can argue a label that is not visible by default is as useful as no label at all. An example scenario is an icon control that has a tooltip and an accName different from that tooltip. In that scenario, because the tooltip is not available by default, it can be argued that there is no "visible label" that would help speech users navigate to that control. Therefore, the criterion can be marked not applicable.

At the same time, another user group intended for 2.5.3 is sighted text to speech users. They would perceive the label, assuming the tooltip is operable for that user group. Therefore, the criterion would be applicable as there is a visible label, and the criterion would fail if the accName was different from that label.

Disclaimer: I am aware that the tooltip is bad.

**Comment by alastc** [comment]
I think you can refer to [3.3.2 Labels or instructions](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html) for this, where the label is defined as "[text](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html#dfn-text) or other component with a [text alternative](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html#dfn-text-alternative) that is presented to a user to identify a component within Web [content](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html#dfn-content)".

We've had previous discussions about persistent labels in #864, and see also #855.

Personally, I think having a non-visible (by default) bit of text wouldn't count as a label.

**Comment by mbgower** [comment]
> Avoiding the rabbit hole of accessible name computation discussed in those threads, the conclusion is that placeholder text cannot be considered a label because it is not persistent.

The [understanding document](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html) directly addresses placeholder text:

> Note that placeholder text within an input field is not considered an appropriate means of providing a label. The [HTML5 specification](https://www.w3.org/TR/html52/sec-forms.html#the-placeholder-attribute) states The placeholder attribute should not be used as an alternative to a <label>. However, it is worth noting that "label" in that HTML5 statement is in code brackets and links to the label element. For the purposes of this Label in Name Success Criterion, "label" is not used in such a programmatic sense but is simply referring to a text string in close visual proximity to a component. As such, in the absence of any other nearby text string (as described in the preceding list), if an input contains placeholder text, such text may be a candidate for Label in Name. This is supported both through the accessible name calculation (discussed later) and from the practical sense that where a visible label is not otherwise provided, it is likely that a speech-input user may attempt to use the placeholder text value as a means of interacting with the input.

In regard to the issue title, I feel like the fairly expansive Understanding document (which also deals with accname computations, btw) addresses what a visible label is, as well as some of the challenges with interpretation.

On your other points, I agree that non-persistent labels of any kind get messy (and may fail other criteria) but in regards to the goal of this SC, where I am operating by speech, if I cannot see a piece of text serving as a label for a control, I obviously can't use the label to reference it. In the case of a placeholder, it is there by default, so I _can_ mentally map it as a label. In the case of a tooltip label, I as a speech operator (without the aid of a mouse), would need to somehow trigger that tooltip to even see it -- at which point I'm already at it, so I'd argue this temporarily displayed info is mostly redundant to the interaction.
For folks who  use a site multiple times AND have good memories, they may be able to benefit from that non-persistent name. So I think one _could_ make a case that a mismatch between an icon-only button's tooltip and its accessible name could potentially cause issues and should be avoided. But I think context is going to come into play here. As is stated in the Understanding document, it pays to be conservative with interpretations of label.

**Comment by anevins12** [comment]
I’m not concerned with placeholder text or AccName computations, so that’s fine. 

It’s true that one of the intended audiences can’t perceive the label prior to interacting with it, but what of the other groups discussed in that criterion, like text to speech users? If the tooltip was perceivable to those users, what would you do then?

**Comment by alastc** [comment]
> It’s true that one of the intended audiences can’t perceive the label prior to interacting with it, but what of the other groups discussed in that criterion

An SC is not required to have all aspects apply to all potential audiences the same way. If the intended audience is impacted, that's enough. 

There is extensive discussion of visible labels, do we need to add an explanation of a niche case of whether a non-visible tooltip is considered visible? Personally, I think anyone who knows the intent and that it's for a "visible label" wouldn't include a hidden-by-default tooltip.
